### PR TITLE
⚡️ `Seek` the `text` section instead of loading the whole `text` section into memory

### LINKED ISSUE #7: Improve Tracer performance, reduce CPU consumption at the scanning stage
**Current**:
- When Tracer starts running it consumes a significant amount of CPU, probably related to the number of processes running on the node as it inspects all processes to find TLS elements (e.g. libraries and GOlang programs)
- The high consumption continues until all processes are inspected.
- This elevated consumption can cause OOMKilled for the Worker pod.
- On OOMKIlled this process begins again.
- While this happens on busy nodes and can be resolved with more resources, we can likely improve this behavior

**Requested**:
- Throttle  the scanning process
- Enable/disable detection of TLS elements (e.g OpenSSL, Golang programs, etc) - this is optional and great for debugging. Yet if you solve the throttling, this feature gains a secondary priority.

### PR DESCRIPTION
**Pull request by iluxa**
Added elf file parsing optimizations

new enf_file.go file is created based on "debug/elf" go module: added scanning symbols support via go channels

resolves #7

### CODE REVIEW

File: go_offsets.go [hunk 1]
```diff
@@ -303,7 +301,19 @@ func getOffsets(fpath string) (offsets map[string]*goExtendedOffset, goidOffset
...
+		}
+		num := int(symEndingIndex - symStartingIndex)
+		var numRead int
+		symBytes := make([]byte, num)
+		numRead, err = textSectionFile.Read(symBytes)
+		if err != nil {
+			return
+		}
+		if numRead != num {
+			err = errors.New("Text section read failed")
```
Comment (mertyildiran) [inline go_offsets.go:315]:
```suggestion
			err = errors.New("Text section read failed")
			return
```

File: go_offsets.go [hunk 2]
```diff
@@ -9,6 +9,8 @@ import (
 	"os"
 	"runtime"
 
+	"errors"
+
```
Comment (mertyildiran) [inline go_offsets.go:13]:
```suggestion
	"runtime"
	"errors"

```

File: elf_file.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	strdata, err := f.stringTable(symtabSection.Link)
+	if err != nil {
+		return nil, fmt.Errorf("cannot load string table section: %w", err)
+	}
+
+	// The first entry is all zeros.
+	var skip [elf.Sym32Size]byte
+	_, _ = symtab.Read(skip[:])
```
Comment (mertyildiran) [inline elf_file.go:63]: Handle error, don't ignore.

File: elf_file.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+	ch := make(chan elf.Symbol, 1)
+
+	var sym elf.Sym32
+	for symtab.Len() > 0 {
+		defer close(ch)
+		_ = binary.Read(symtab, f.ByteOrder, &sym)
```
Comment (mertyildiran) [inline elf_file.go:70]: Handle error, don't ignore.

File: elf_file.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+		str, _ := getString(strdata, int(sym.Name))
```
Comment (mertyildiran) [inline elf_file.go:71]: Handle error, don't ignore.

File: elf_file.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+	// The first entry is all zeros.
+	var skip [elf.Sym64Size]byte
+	_, _ = symtab.Read(skip[:])
+
+	ch := make(chan elf.Symbol, 1)
+	go func() {
+		defer close(ch)
+		var sym elf.Sym64
+		for symtab.Len() > 0 {
+			_ = binary.Read(symtab, f.ByteOrder, &sym)
```
Comment (mertyildiran) [inline elf_file.go:114]: Handle error, don't ignore.

File: elf_file.go [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	strdata, err := f.stringTable(symtabSection.Link)
+	if err != nil {
+		return nil, fmt.Errorf("cannot load string table section: %w", err)
+	}
+
+	// The first entry is all zeros.
+	var skip [elf.Sym64Size]byte
+	_, _ = symtab.Read(skip[:])
```
Comment (mertyildiran) [inline elf_file.go:107]: Handle error, don't ignore.

File: elf_file.go [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+	ch := make(chan elf.Symbol, 1)
+	go func() {
+		defer close(ch)
+		var sym elf.Sym64
+		for symtab.Len() > 0 {
+			_ = binary.Read(symtab, f.ByteOrder, &sym)
+			str, _ := getString(strdata, int(sym.Name))
```
Comment (mertyildiran) [inline elf_file.go:115]: Handle error, don't ignore.

File: elf_file.go [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+type ElfFile struct {
+	*elf.File
+}
+
+func NewElfFile(r io.ReaderAt) (*ElfFile, error) {
+	f, err := elf.NewFile(r)
+	ret := &ElfFile{
+		File: f,
+	}
+	return ret, err
```
Comment (mertyildiran) [inline elf_file.go:21]:
```suggestion
	f, err := elf.NewFile(r)
	if err != nil {
		return nil, err	
	}
	return &ElfFile{
		File: f,
	}, nil
```

File: elf_file.go [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+func (f *ElfFile) SymbolsChan() (chan elf.Symbol, error) {
+	sym, err := f.getSymbolsChan(elf.SHT_SYMTAB)
+	return sym, err
```
Comment (mertyildiran) [inline elf_file.go:26]:
```suggestion
	return f.getSymbolsChan(elf.SHT_SYMTAB)
```

**Review by mertyildiran — CHANGES_REQUESTED** [review]

File: elf_file.go (hunk 3 above)
Comment (iluxa) [reply elf_file.go]: removed file

File: elf_file.go (hunk 4 above)
Comment (iluxa) [reply elf_file.go]: removed file

File: elf_file.go (hunk 6 above)
Comment (iluxa) [reply elf_file.go]: removed file

File: elf_file.go (hunk 8 above)
Comment (iluxa) [reply elf_file.go]: removed file

File: elf_file.go (hunk 9 above)
Comment (iluxa) [reply elf_file.go]: removed file

File: elf_file.go (hunk 10 above)
Comment (iluxa) [reply elf_file.go]: removed file

**Pull request merged**